app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php

```diff
@@ -108,7 +118,9 @@ public function addProducts($productIds)
     {
         if (is_array($productIds)) {
             foreach ($productIds as $productId) {
-                $this->addProduct($productId);
+                if ($this->productRepository->getById($productId)->getId()) {
```

Comment:
This seems to be not effective. Please obtain all products at once with `SearchCriteria` or something.

Reply:
used SearchCriteriaBuilder